File: pylox/pylox/Scanner.py
from typing import Callable, Final
from Token import Token
from TokenTypes import TokenType


class Scanner:
    keywords = {
        "and": TokenType.AND,
        "class": TokenType.CLASS,
        "else": TokenType.ELSE,
        "false": TokenType.FALSE,
        "for": TokenType.FOR,
        "fun": TokenType.FUN,
        "if": TokenType.IF,
        "nil": TokenType.NIL,
        "or": TokenType.OR,
        "print": TokenType.PRINT,
        "return": TokenType.RETURN,
        "super": TokenType.SUPER,
        "this": TokenType.THIS,
        "true": TokenType.TRUE,
        "var": TokenType.VAR,
        "while": TokenType.WHILE,
    }

    def __init__(self, source: str, error: Callable[[int, str], None]) -> None:
        self._source: Final[str] = source
        self._tokens: Final[list[Token]] = []
        self._start = 0
        self._current = 0
        self._line = 1
        self._error: Final[Callable[[int, str], None]] = error

    def scan_tokens(self) -> list[Token]:
        while not self._is_at_end():
            self._start = self._current
            self._scan_token()

        self._tokens.append(Token(TokenType.EOF, "", None, self._line))
        return self._tokens

    def _is_at_end(self) -> bool:
        return self._current >= len(self._source)

    def _scan_token(self) -> None:
        c: str = self._advance()
        match c:
            case "(":
                self._add_token(TokenType.LEFT_PAREN)
            case ")":
                self._add_token(TokenType.RIGHT_PAREN)
            case "{":
                self._add_token(TokenType.LEFT_BRACE)
            case "}":
                self._add_token(TokenType.RIGHT_BRACE)
            case ",":
                self._add_token(TokenType.COMMA)
            case ".":
                self._add_token(TokenType.DOT)
            case "-":
                self._add_token(TokenType.MINUS)
            case "+":
                self._add_token(TokenType.PLUS)
            case ";":
                self._add_token(TokenType.SEMICOLON)
            case "*":
                self._add_token(TokenType.STAR)
            case "!":
                self._add_token(
                    TokenType.BANG_EQUAL if self._match("=") else TokenType.BANG
                )
            case ">":
                self._add_token(
                    TokenType.GREATER_EQUAL if self._match("=") else TokenType.GREATER
                )
            case "<":
                self._add_token(
                    TokenType.LESS_EQUAL if self._match("=") else TokenType.LESS
                )
            case "=":
                self._add_token(
                    TokenType.EQUAL_EQUAL if self._match("=") else TokenType.EQUAL
                )
            case "/":
                if self._match("/"):
                    while self._peek() != "\n" and not self._is_at_end():
                        self._advance()
                else:
                    self._add_token(TokenType.SLASH)
            case " " | "\r" | "\t":
                pass
            case "\n":
                self._line += 1
            case '"':
                self._string()
            case _:
                if self._is_digit(c):
                    self._number()
                elif self._is_alpha(c):
                    self._identifier()
                else:
                    self._error(self._line, "Unexpected character.")

    def _add_token(self, type: TokenType, literal: object = None) -> None:
        lexeme = self._source[self._start : self._current]
        self._tokens.append(Token(type, lexeme, literal, self._line))

    def _advance(self) -> str:
        c: str = self._source[self._current]
        self._current += 1
        return c

    def _peek(self) -> str:
        if self._is_at_end():
            return "\0"
        return self._source[self._current]

    def _peek_next(self) -> str:
        if self._current + 1 >= len(self._source):
            return "\0"
        return self._source[self._current + 1]

    def _match(self, expected: str) -> bool:
        if self._is_at_end():
            return False
        if self._source[self._current] != expected:
            return False

        self._current += 1
        return True

    def _string(self) -> None:
        while self._peek() != '"' and not self._is_at_end():
            if self._peek() == "\n":
                self._line += 1
            self._advance()

        if self._is_at_end():
            self._error(self._line, "Unterminated string.")
            return

        # The closing quote
        self._advance()

        # Trim the surrounding quotes
        value = self._source[self._start + 1 : self._current - 1]

        self._add_token(TokenType.STRING, value)

    def _is_digit(self, c: str) -> bool:
        return len(c) == 1 and c.isdigit()

    def _is_alpha(self, c: str) -> bool:
        return len(c) == 1 and (c.isalpha() or c == "_")

    def _is_alpha_numeric(self, c: str) -> bool:
        return self._is_alpha(c) or self._is_digit(c)

    def _number(self) -> None:
        while self._is_digit(self._peek()):
            self._advance()

        # Look for a fractional part
        if self._peek() == "." and self._is_digit(self._peek_next()):
            # Consume the '.'
            self._advance()
            while self._peek().isdigit():
                self._advance()

        value = float(self._source[self._start : self._current])

        self._add_token(TokenType.NUMBER, value)

    def _identifier(self):
        while self._is_alpha_numeric(self._peek()):
            self._advance()

        text = self._source[self._start : self._current]

        type = Scanner.keywords.get(text)

        if not type:
            type = TokenType.IDENTIFIER

        self._add_token(type)

File: golox/internal/scanner/scanner.go
package scanner

import (
	"fmt"
	"log"
	"strconv"
	"unicode"

	"github.com/englishlayup/crafting_interpreters/golox/internal/token"
)


type Scanner struct {
	source  string
	tokens  []token.Token
	current int
	start   int
	line    int
	report  func(int, string)
}

func NewScanner(source string, report func(int, string)) *Scanner {
	return &Scanner{
		source:  source,
		tokens:  make([]token.Token, 0),
		current: 0,
		start:   0,
		line:    1,
		report:  report,
	}
}

func (s *Scanner) ScanTokens() []token.Token {
	for !s.isAtEnd() {
		s.start = s.current
		s.scanToken()
	}

	s.tokens = append(s.tokens, *token.NewToken(token.EOF, "", nil, s.line))
    return s.tokens
}

func (s *Scanner) scanToken() {
	c := s.advance()

	switch c {
	case '(':
		s.addToken(token.LEFT_PAREN, nil)
	case ')':
		s.addToken(token.RIGHT_PAREN, nil)
	case '{':
		s.addToken(token.LEFT_BRACE, nil)
	case '}':
		s.addToken(token.RIGHT_BRACE, nil)
	case ',':
		s.addToken(token.COMMA, nil)
	case '.':
		s.addToken(token.DOT, nil)
	case '-':
		s.addToken(token.MINUS, nil)
	case '+':
		s.addToken(token.PLUS, nil)
	case ';':
		s.addToken(token.SEMICOLON, nil)
	case '*':
		s.addToken(token.STAR, nil)
	case '!':
		if s.match('=') {
			s.addToken(token.BANG_EQUAL, nil)
		} else {
			s.addToken(token.BANG, nil)
		}
	case '>':
		if s.match('=') {
			s.addToken(token.GREATER_EQUAL, nil)
		} else {
			s.addToken(token.GREATER, nil)
		}
	case '<':
		if s.match('=') {
			s.addToken(token.LESS_EQUAL, nil)
		} else {
			s.addToken(token.LESS, nil)
		}
	case '=':
		if s.match('=') {
			s.addToken(token.EQUAL_EQUAL, nil)
		} else {
			s.addToken(token.EQUAL, nil)
		}
	case '/':
		if s.match('/') {
			for s.peek() != '\n' && !s.isAtEnd() {
				s.advance()
			}
		} else {
			s.addToken(token.SLASH, nil)
		}
	case ' ', '\r', '\t':
	case '\n':
		s.line += 1
	case '"':
		s.scanString()
	default:
        if s.isDigit(c) {
            s.scanNumber()
        } else if s.isAlpha(c) {
            s.identifier()
        } else {
            s.report(s.line, "Unexpected character.")
        }
	}
}

func (s *Scanner) isAlphaNumeric(c byte) bool {
    return s.isAlpha(c) || s.isDigit(c)
}

func (s *Scanner) isAlpha(c byte) bool {
    return unicode.IsLetter(rune(c)) || c == '_'
}

func (s *Scanner) isDigit(c byte) bool {
    return unicode.IsDigit(rune(c))
}

func (s *Scanner) identifier() {
    var keywords = map[string]token.TokenType{
        "and":    token.AND,
        "class":  token.CLASS,
        "else":   token.ELSE,
        "false":  token.FALSE,
        "for":    token.FOR,
        "fun":    token.FUN,
        "if":     token.IF,
        "nil":    token.NIL,
        "or":     token.OR,
        "print":  token.PRINT,
        "return": token.RETURN,
        "super":  token.SUPER,
        "this":   token.THIS,
        "true":   token.TRUE,
        "var":    token.VAR,
        "while":  token.WHILE,
    }

    for s.isAlphaNumeric(s.peek()){
        s.advance()
    }

    text := s.source[s.start:s.current]
    tokenType, exist := keywords[text]

    if !exist {
        tokenType = token.IDENTIFIER
    }

    s.addToken(tokenType, nil)
}

func (s *Scanner) scanNumber() {
    for s.isDigit(s.peek()){
        s.advance()
    }

    if s.peek() == '.' && s.isDigit(s.peekNext()) {
        s.advance()
        for s.isDigit(s.peek()){
            s.advance()
        }
    }

    value, err := strconv.ParseFloat(s.source[s.start:s.current], 64)
    if err != nil {
        log.Fatal(err)
    }
    s.addToken(token.NUMBER, value)
}

func (s *Scanner) scanString() {
	for s.peek() != '"' && !s.isAtEnd() {
		if s.peek() == '\n' {
			s.line += 1
		}
		s.advance()
	}

	if s.isAtEnd() {
		s.report(s.line, "Unterminated string.")
		return
	}

	s.advance()
	value := s.source[s.start+1 : s.current-1]
	s.addToken(token.STRING, value)
}

func (s *Scanner) peek() byte {
    if s.isAtEnd() {
        return 0
    }
	return s.source[s.current]
}

func (s *Scanner) peekNext() byte {
    if s.current + 1 >= len(s.source) {
        return 0
    }
	return s.source[s.current+1]
}

func (s *Scanner) match(expected byte) bool {
	if s.isAtEnd() {
		return false
	}
	if expected != s.source[s.current] {
		return false
	}

	s.current += 1
	return true
}

func (s *Scanner) addToken(tokenType token.TokenType, literal interface{}) {
	lexeme := s.source[s.start:s.current]
	s.tokens = append(s.tokens, *token.NewToken(tokenType, lexeme, literal, s.line))
}

func (s *Scanner) advance() byte {
	c := s.source[s.current]
	s.current += 1
	return c
}

func (s *Scanner) isAtEnd() bool {
	return s.current == len(s.source)-1
}
